File: sdk/python/examples/outlier_kill.py
"""Watch a workflow's live iteration count and kill it if it runs away.

Run the gateway in loop mode to see this fire:
    ECHO_LOOP=true docker compose up -d agent-bus-app
"""
import asyncio

from agent_bus_client import AgentBusClient

GATEWAY = "http://127.0.0.1:6815"
STEP_BUDGET = 20


async def main() -> None:
    async with AgentBusClient(GATEWAY) as client:
        wf = await client.start("loop forever please")
        print("workflow:", wf.cid, "| budget:", STEP_BUDGET)

        async for ev in wf:
            print(f"  sid={ev.sid:>3}  {ev.type}")
            if ev.sid >= STEP_BUDGET and ev.type != "workflow.terminated":
                snap = await wf.status()
                print("  budget exceeded:", snap, "-> terminating")
                await wf.terminate()

        print("final status:", await wf.status())


if __name__ == "__main__":
    asyncio.run(main())

File: sdk/javascript/examples/outlier-kill.mjs
// Watch a workflow's live iteration count and kill it if it runs away.
//   ECHO_LOOP=true docker compose up -d agent-bus-app
//   node examples/outlier-kill.mjs
import { AgentBusClient } from "../src/agent-bus-client.js";

const GATEWAY = "http://127.0.0.1:6815";
const STEP_BUDGET = 20;

const client = new AgentBusClient(GATEWAY);
await client.connect();

const wf = await client.start("loop forever please");
console.log("workflow:", wf.cid, "| budget:", STEP_BUDGET);

for await (const ev of wf) {
  console.log(`  sid=${String(ev.sid).padStart(3)}  ${ev.type}`);
  if (ev.sid >= STEP_BUDGET && !ev.isTerminal) {
    console.log("  budget exceeded:", await wf.status(), "-> terminating");
    await wf.terminate();
  }
}

console.log("final status:", await wf.status());
client.disconnect();
